docs/source/java/memory.rst

```diff
@@ -0,0 +1,208 @@
+``ReferenceManager.retain``.
+
+Of course, this is tedious and error-prone, so usually, instead of directly working with buffers, we should use
+higher-level APIs like ValueVector. Such classes generally implement Closeable/AutoCloseable and will automatically
+decrement the reference count when closed.
+
+Allocators implement AutoCloseable as well. In this case, closing the allocator will check that all buffers
+obtained from the allocator are closed. If not, ``close()`` method will raise an exception; this helps track
+memory leaks from unclosed buffers.
+
+As you see, reference counting needs to be handled carefully. To ensure that an
+independent section of code has fully cleaned up all allocated buffers, use a new child allocator.
+
+Development Guidelines
+======================
+
+Applications should generally:
+
+* Use the BufferAllocator interface in APIs instead of RootAllocator.
+* Create one RootAllocator at the start of the program.
+* ``close()`` allocators after use (whether they are child allocators or the RootAllocator), either manually or preferably via a try-with-resources statement.
+
+Debugging Memory Leaks/Allocation
+=================================
+
+Allocators have a debug mode that makes it easier to figure out where a leak is originated.
+To enable it, enable assertions with ``-ea`` or set the system property, ``-Darrow.memory.debug.allocator=true``.
+When enabled, a log will be kept of allocations.
+
+Arrow modules define simple logging facade for java SLF4J, configure it properly to see your logs (e.g. Logback/Log4J).
+
+Consider the following example to see how debug enabled help us with the tracking of allocators:
+
+.. code-block:: Java
+
+    import org.apache.arrow.memory.BufferAllocator;
+    import org.apache.arrow.memory.RootAllocator;
+    import org.apache.arrow.vector.IntVector;
+
+    try (BufferAllocator bufferAllocator = new RootAllocator(Integer.MAX_VALUE)) {
+        final int QUANTITY = 5;
+        try (IntVector intVector = new IntVector("int-01", bufferAllocator)) {
+            intVector.allocateNew(QUANTITY);
+            for (int i = 0; i < QUANTITY; i++) {
+                intVector.set(i, i);
+            }
+            intVector.setValueCount(QUANTITY);
+        }
+        // Fix the next code!, it is only to see the track of allocators when debug is enabled
+        IntVector intVectorV = new IntVector("int-02", bufferAllocator);
+        intVectorV.allocateNew(QUANTITY);
+        for (int i = 0; i < QUANTITY; i++) {
+            intVectorV.set(i, i);
+        }
+        intVectorV.setValueCount(QUANTITY);
+
+        BufferAllocator childAllocator = bufferAllocator.newChildAllocator("child-isolated", 0,
+                Integer.MAX_VALUE / 4);
+        IntVector intVectorV2 = new IntVector("int-isolated-01", childAllocator);
+        intVectorV2.allocateNew(QUANTITY);
+        for (int i = 0; i < QUANTITY; i++) {
+            intVectorV2.set(i, i);
+        }
+    }
```

Comment:
We should only use ArrowBuf here, don't use IntVector or anything as they haven't been introduced yet.

Also, it's still not clear to me what the debug mode is doing here. From what I recall, the information it provides is basically only available in a debugger. If there is something the debug mode is doing here, we need to point it out explicitly. Ideally we would compare the information available with and without debug mode.

Reply:
Added a example with/without use this parameter

Reply:
Thanks.

I'm pretty sure you can get a stack trace showing where each buffer was allocated, but that may only be in the debugger. I'll try to follow up tomorrow…

Reply:
I'll submit a jira/PR for this later (there's some issues that need fixing) but I think things are good here as is.